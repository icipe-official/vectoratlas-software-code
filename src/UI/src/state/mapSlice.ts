import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { fetchApiJson } from '../api/api';

export interface MapState {
  map_styles: {
    layers: {
      name: string;
      fillColor: number[];
      strokeColor: number[];
      strokeWidth: number;
      zIndex: number;
    }[];
  };

  map_overlays: {
    name: string;
    source: string;
    sourceType: string;
    layers: { name: string }[];
  }[];
}

export const initialState: MapState = {
  map_styles: { layers: [] },
  map_overlays: [],
};

export const getMapStyles = createAsyncThunk('map/getMapStyles', async () => {
  const mapStyles = await fetchApiJson('config/map-styles');
  return mapStyles;
});

export const getTileServerOverlays = createAsyncThunk(
  'map/getTileServerOverlays',
  async () => {
    const tileServerOverlays = await fetchApiJson(
      'config/tile-server-overlays'
    );
    return tileServerOverlays;
  }
);

export const mapSlice = createSlice({
  name: 'map',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(getMapStyles.fulfilled, (state, action) => {
        state.map_styles = action.payload;
      })
      .addCase(getTileServerOverlays.fulfilled, (state, action) => {
        state.map_overlays = action.payload;
      });
  },
});

export default mapSlice.reducer;
