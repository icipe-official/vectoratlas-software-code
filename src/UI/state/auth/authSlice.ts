import { createSlice } from '@reduxjs/toolkit';
import { getUserInfo } from './actions/getUserInfo';

export interface AuthState {
  roles: String[];
  token: String;
  isLoading: Boolean;
  roleRequestLoading: boolean;
}

export const initialState: () => AuthState = () => ({
  roles: [],
  token: '',
  isLoading: true,
  roleRequestLoading: false,
});

export const authSlice = createSlice({
  name: 'auth',
  initialState: initialState(),
  reducers: {
    requestLoading(state, action) {
      state.roleRequestLoading = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(getUserInfo.pending, (state) => {
      state.isLoading = true;
    });
    builder.addCase(getUserInfo.rejected, (state) => {
      state.isLoading = false;
    });
    builder.addCase(getUserInfo.fulfilled, (state, action: any) => {
      state.roles = action.payload.roles;
      state.token = action.payload.token;
      state.isLoading = false;
    });
  },
});

export const { requestLoading } = authSlice.actions;

export default authSlice.reducer;
