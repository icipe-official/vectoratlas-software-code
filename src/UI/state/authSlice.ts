import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import * as njwt from 'njwt';
import { fetchAuth } from '../api/api';

export interface AuthState {
  roles: String[];
  token: String;
}

export const initialState: AuthState = {
  roles: [],
  token: '',
};

export const getUserInfo = createAsyncThunk('auth/getUserInfo', async () => {
  const token = await fetchAuth();
  const verifiedToken: any = njwt.verify(
    token,
    process.env.NEXT_PUBLIC_TOKEN_KEY
  );
  return {
    token: token,
    roles: verifiedToken?.body.scope.split(','),
  };
});

export const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(getUserInfo.fulfilled, (state, action: any) => {
      state.roles = action.payload.roles;
      state.token = action.payload.token;
    });
  },
});

export default authSlice.reducer;
