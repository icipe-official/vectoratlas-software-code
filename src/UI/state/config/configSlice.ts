import { createSlice } from '@reduxjs/toolkit';
import { getAllData } from './actions/getAllData';
import { getApiVersion } from './actions/getApiVersion';
import { getFeatureFlags } from './actions/getFeatureFlags';
import { getUiVersion } from './actions/getUiVersion';

export interface ConfigState {
  version_ui: string;
  version_api: string;
  feature_flags_status: string;
  download_all_status: string;
  feature_flags: {
    flag: string;
    on: boolean;
  }[];
}

export const initialState: ConfigState = {
  version_ui: 'local_ui',
  version_api: 'local_api',
  feature_flags: [],
  feature_flags_status: '',
  download_all_status: '',
};

export const configSlice = createSlice({
  name: 'config',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(getUiVersion.pending, (state) => {
        state.version_ui = 'loading';
      })
      .addCase(getUiVersion.rejected, (state) => {
        state.version_ui = 'error';
      })
      .addCase(getUiVersion.fulfilled, (state, action) => {
        state.version_ui = action.payload;
      })
      .addCase(getApiVersion.pending, (state) => {
        state.version_api = 'loading';
      })
      .addCase(getApiVersion.rejected, (state) => {
        state.version_api = 'error';
      })
      .addCase(getApiVersion.fulfilled, (state, action) => {
        state.version_api = action.payload;
      })
      .addCase(getFeatureFlags.pending, (state) => {
        state.feature_flags_status = 'loading';
      })
      .addCase(getFeatureFlags.rejected, (state) => {
        state.feature_flags_status = 'error';
      })
      .addCase(getFeatureFlags.fulfilled, (state, action) => {
        state.feature_flags_status = 'success';
        state.feature_flags = action.payload;
      })
      .addCase(getAllData.fulfilled, (state) => {
        state.download_all_status = 'success';
      })
      .addCase(getAllData.rejected, (state) => {
        state.download_all_status = 'error';
      });
  },
});

export default configSlice.reducer;
