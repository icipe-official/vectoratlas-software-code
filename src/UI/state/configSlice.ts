import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import {
  fetchLocalVersion,
  fetchApiVersion,
  fetchFeatureFlags,
} from '../api/api';

export interface ConfigState {
  version_ui: string;
  version_api: string;
  feature_flags_status: string;
  feature_flags: {
    flag: string;
    on: boolean;
  }[];
}

export const initialState: ConfigState = {
  version_ui: 'local_ui',
  version_api: 'local_api',
  feature_flags: [],
  feature_flags_status: '',
};

export const getUiVersion = createAsyncThunk(
  'config/getUiVersion',
  async () => {
    const version = await fetchLocalVersion();
    return version;
  }
);

export const getApiVersion = createAsyncThunk(
  'config/getApiVersion',
  async () => {
    const version = await fetchApiVersion();
    return version;
  }
);

export const getFeatureFlags = createAsyncThunk(
  'config/getFeatureFlags',
  async () => {
    const featureFlags = await fetchFeatureFlags();
    return featureFlags;
  }
);

export const configSlice = createSlice({
  name: 'config',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(getUiVersion.pending, (state) => {
        state.version_ui = 'loading';
      })
      .addCase(getUiVersion.rejected, (state) => {
        state.version_ui = 'error';
      })
      .addCase(getUiVersion.fulfilled, (state, action) => {
        state.version_ui = action.payload;
      })
      .addCase(getApiVersion.pending, (state) => {
        state.version_api = 'loading';
      })
      .addCase(getApiVersion.rejected, (state) => {
        state.version_api = 'error';
      })
      .addCase(getApiVersion.fulfilled, (state, action) => {
        state.version_api = action.payload;
      })
      .addCase(getFeatureFlags.pending, (state) => {
        state.feature_flags_status = 'loading';
      })
      .addCase(getFeatureFlags.rejected, (state) => {
        state.feature_flags_status = 'error';
      })
      .addCase(getFeatureFlags.fulfilled, (state, action) => {
        state.feature_flags_status = 'success';
        state.feature_flags = action.payload;
      });
  },
});

export default configSlice.reducer;
