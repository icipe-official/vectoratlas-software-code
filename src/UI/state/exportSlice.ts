// import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
// import { fetchApiJson } from '../api/api';

// export const getAll = createAsyncThunk('export/getAllOccurrenceData', async () => {
//     const allOccurrenceData = await fetchApiJson('export/stream-all-file');
//     return allOccurrenceData;
//   }
// );

// export const initialState: any = []

// export const mapSlice = createSlice({
//   name: 'map',
//   initialState
//   reducers: {}
    
//   extraReducers: (builder) => {
//     builder
//       .addCase(getMapStyles.fulfilled, (state, action) => {
//         state.map_styles = action.payload;
//       })
//       .addCase(getTileServerOverlays.fulfilled, (state, action) => {
//         state.map_overlays = action.payload;
//       });

//   },
// });

// export const {
  
// } = mapSlice.actions;

// export default mapSlice.reducer;
