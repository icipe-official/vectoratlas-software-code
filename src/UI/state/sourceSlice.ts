import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import { fetchGraphQlData, fetchGraphQlDataAuthenticated } from "../api/api";
import { sourceQuery, newSourceQuery } from "../api/queries";
import { NewSource } from "../components/sources/source_form";
import { AppState } from "./store";


export interface SourceState {
  source_info: {
      author: string,
      article_title:string,
      journal_title:string,
      citation: string,
      year: number,
      published:boolean,
      report_type: string,
      v_data: boolean

  }[],
  source_info_status: string,
}

export const initialState: SourceState = {
source_info: [],
source_info_status: "",

}

//Genereting pending, fulfilled and rejected action types
export const getSourceInfo = createAsyncThunk('source/getSourceInfo', async () => {
  const sourceInfo = await fetchGraphQlData(sourceQuery());

  return sourceInfo.data.allReferenceData;
})

export const postNewSource = createAsyncThunk('source/getSourceInfo', async (source: NewSource, { getState }) => {
  const query = newSourceQuery(source);
  const token = (getState() as AppState).auth.token
  const queryResponse = await fetchGraphQlDataAuthenticated(query, token);
  console.log(queryResponse);
})

export const sourceSlice = createSlice({
  name: 'source_info',
  initialState,
  reducers:{
  },
  extraReducers: (builder) => {
    builder
    .addCase(getSourceInfo.pending, (state) => {
      state.source_info_status = 'loading';
    })
    .addCase(getSourceInfo.rejected, (state, action) => {
      state.source_info_status = 'error';
    })
    .addCase(getSourceInfo.fulfilled, (state, action) => {
      state.source_info = action.payload;
      state.source_info_status = 'success';
    })

  },
})

export default sourceSlice.reducer;