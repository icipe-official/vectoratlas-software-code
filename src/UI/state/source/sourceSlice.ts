import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { FilterSort } from '../state.types';
import { getSourceInfo } from './actions/getSourceInfo';

export interface Source {
  [index: string]: any;
  author: string;
  article_title: string;
  journal_title: string;
  citation: string;
  year: number;
  published: boolean;
  report_type: string;
  v_data: boolean;
  num_id: number;
}

export interface SourceState {
  source_info: {
    items: Source[];
    total: number;
  };
  source_info_status: string;
  source_table_options: FilterSort;
}

export const initialState: SourceState = {
  source_info: {
    items: [],
    total: 0,
  },
  source_info_status: '',
  source_table_options: {
    page: 0,
    rowsPerPage: 10,
    orderBy: 'num_id',
    order: 'asc',
    startId: 0,
    endId: null,
    textFilter: '',
  },
};

export const sourceSlice = createSlice({
  name: 'source_info',
  initialState,
  reducers: {
    changeSourcePage(state, action: PayloadAction<number>) {
      state.source_table_options.page = action.payload;
    },
    changeSourceRowsPerPage(state, action: PayloadAction<number>) {
      state.source_table_options.rowsPerPage = action.payload;
    },
    changeSort(state, action: PayloadAction<string>) {
      const isAsc =
        state.source_table_options.orderBy === action.payload &&
        state.source_table_options.order === 'asc';
      state.source_table_options.order = isAsc ? 'desc' : 'asc';
      state.source_table_options.orderBy = action.payload;
    },
    changeFilterId(
      state,
      action: PayloadAction<{ startId: number | null; endId: number | null }>
    ) {
      state.source_table_options.startId = action.payload.startId;
      state.source_table_options.endId = action.payload.endId;
    },
    changeFilterText(state, action: PayloadAction<string>) {
      state.source_table_options.textFilter = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getSourceInfo.pending, (state) => {
        state.source_info_status = 'loading';
      })
      .addCase(getSourceInfo.rejected, (state, action) => {
        state.source_info_status = 'error';
      })
      .addCase(getSourceInfo.fulfilled, (state, action) => {
        state.source_info = action.payload;
        state.source_info_status = 'success';
      });
  },
});

export const {
  changeSourcePage,
  changeSourceRowsPerPage,
  changeSort,
  changeFilterId,
  changeFilterText,
} = sourceSlice.actions;
export default sourceSlice.reducer;
