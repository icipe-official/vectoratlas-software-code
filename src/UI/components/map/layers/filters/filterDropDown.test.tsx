import { CombinedState } from '@reduxjs/toolkit';
import React from 'react';
import { AuthState } from '../../../../state/authSlice';
import { ConfigState } from '../../../../state/configSlice';
import {
  filterHandler,
  initialState,
  MapState,
} from '../../../../state/map/mapSlice';
import { SourceState } from '../../../../state/sourceSlice';
import { fireEvent, render, within } from '../../../../test_config/render';
import FilterDropDown from './filterDropDown';

/* eslint-disable react/display-name*/
jest.mock('@mui/material/Autocomplete', () => (props) => (
  <div>
    <div>Autocomplete mock {JSON.stringify(props)}</div>
    <input role="input" onChange={(e) => props.onChange(e, e.target.value)} />
  </div>
));
/* eslint-enable react/display-name*/

describe('FilterDropDown', () => {
  let state:
    | Partial<
        CombinedState<{
          config: ConfigState;
          map: MapState;
          auth: AuthState;
          source: SourceState;
        }>
      >
    | undefined;

  beforeEach(() => {
    state = { map: initialState() };
  });

  it('renders correctly for a given filter', () => {
    state.map.filterValues.country = ['Kenya', 'Uganda', 'Tanzania'];
    state.map.filters.country.value = ['Kenya'];

    const { wrapper } = render(
      <FilterDropDown filterTitle={'Country'} filterName="country" />,
      state
    );

    expect(wrapper.container).toMatchSnapshot();
  });

  it('dispatches the correct filter update when the value changes', async () => {
    const { wrapper, store } = render(
      <FilterDropDown filterTitle={'Country'} filterName="country" />,
      state
    );

    const input = await wrapper.getByRole('input');
    fireEvent.change(input, { target: { value: 'Kenya' } });

    expect(store.getActions()[0]).toEqual(
      filterHandler({
        filterName: 'country',
        filterOptions: 'Kenya',
      })
    );
  });
});
