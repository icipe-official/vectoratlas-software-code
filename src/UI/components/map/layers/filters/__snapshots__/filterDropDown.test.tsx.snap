// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterDropDown renders correctly for a given filter 1`] = `
<div>
  <div
    style="display: flex; flex-direction: column; margin-top: 5px; margin-bottom: 10px;"
  >
    <p
      class="MuiTypography-root MuiTypography-inherit css-99b57d-MuiTypography-root"
    >
      Country
    </p>
    <div>
      <div>
        Autocomplete mock 
        {"multiple":true,"size":"small","options":["Kenya","Uganda","Tanzania"],"disableCloseOnSelect":true,"style":{"width":"100%"}}
      </div>
      <input
        role="input"
      />
    </div>
  </div>
</div>
`;
